coffee-chats/src/main/java/com/google/step/coffee/data/RequestStore.java

```diff
@@ -121,4 +126,39 @@ private ChatRequest getRequestFromEntity(Entity entity) throws InvalidEntityExce
 
     return request;
   }
+
+  /**
+   * Removes an existing ChatRequest from datastore and migrates it to an expired request (i.e. no
+   * matches were found in time).
+   */
+  public void expiredRequest(ChatRequest request) {
+    Key expiredKey = KeyFactory.createKey("ChatRequest", request.getRequestId());
+    Entity expiredEntity = createChatRequestEntity(request, "ExpiredChatRequest");
+
+    datastore.put(expiredEntity);
```

Comment:
If server crashes after this line, datastore will have both expired entry and non-expired entry. Then, when this function is called again, this line will try to put it to datastore again - will it throw any exception?

Reply:
As we are using the same key each time for the expired entry, when we call `datastore.put(expiredEntity)` it performs an upsert (see [here](https://cloud.google.com/datastore/docs/concepts/entities#creating_an_entity)) which will overwrite any existing entry if it already exists. So there shouldn't be any duplicates or exceptions raised.